Ensure TrackingPreventionEnabled flag is correctly set before using WebsiteDataStore

szewai opened the pull request:
#### 64fda2fa68e045b76c91af8864e0cde1d8ffe896
<pre>
Ensure TrackingPreventionEnabled flag is correctly set before using WebsiteDataStore
<a href="https://bugs.webkit.org/show_bug.cgi?id=270382">https://bugs.webkit.org/show_bug.cgi?id=270382</a>
<a href="https://rdar.apple.com/123922825">rdar://123922825</a>

Reviewed by Chris Dumez.

We currently set TrackingPreventionEnabled on WebsiteDataStore in WebProcessPool::createNewWebProcess. This has caused
problem that TrackingPreventionEnabled is incorrectly set when WebsiteDataStore (or its corresponding NetworkSession)
is in use. One case is when process is created for prewarm, it does not have a attached WebsiteDataStore (it will be
assigned a WebsiteDataStore later), so we don&apos;t set TrackingPreventionEnabled flag, and we don&apos;t set it when the
prewarmed process is used for loading (WebsiteDataStore is assigned to it). Another case is when WebsiteDataStore is
used without web process for website data retrieval. Failing to set the flag will cause ITP data not to be fetched or
removed.

To fix this issue, this patch moves TrackingPreventionEnabled setting to WebsiteDataStore: WebsiteDataStore will get
the right value when it needs to send its parameters to network process and web process (i.e. it&apos;s going to be used).
To simply the setting logic, this patch also makes the following changes:
1. WebsiteDataStore::m_trackingPreventionEnabled now has 3 states: yes, no and default, where default means we will fall
back to TCC permission state to decide the value. This enables us to remove m_isTrackingPreventionStateExplicitlySet.
2. Drop appHasRequestedCrossWebsiteTrackingPermission and TrackingPreventionStateExplicitlySet parameters since network
process does not need to re-compute the TrackingPreventionEnabled value and know about how it is computed. We previously
used it for logging and debugging purpose, and we probably don&apos;t need it any more (as TCC check is not free). This patch
replaces it with normal logging in NetworkSession::setTrackingPreventionEnabled.

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::setTrackingPreventionEnabled):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.serialization.in:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::NetworkSessionCocoa):
(WebKit::activateSessionCleanup): Deleted.
* Source/WebKit/Shared/ResourceLoadStatisticsParameters.h:
* Source/WebKit/Shared/ResourceLoadStatisticsParameters.serialization.in:
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.cpp:
(WKWebsiteDataStoreSetResourceLoadStatisticsEnabled):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _setResourceLoadStatisticsEnabled:]):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createNewWebProcess):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::WebsiteDataStore):
(WebKit::WebsiteDataStore::setIsRunningResourceLoadStatisticsTest):
(WebKit::WebsiteDataStore::defaultTrackingPreventionEnabled const):
(WebKit::WebsiteDataStore::trackingPreventionEnabled const):
(WebKit::WebsiteDataStore::setTrackingPreventionEnabled):
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::isTrackingPreventionStateExplicitlySet const): Deleted.
(WebKit::WebsiteDataStore::useExplicitTrackingPreventionState): Deleted.

Canonical link: <a href="https://commits.webkit.org/275635@main">https://commits.webkit.org/275635@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0e39d7816173100f39910fba2ff5707b082b5210

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42373 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21391 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44767 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44972 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38490 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24615 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18736 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42947 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18326 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/36487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16038 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46437 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/38568 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37861 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41774 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17189 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14162 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/40376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18808 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9480 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18870 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18453 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/NetworkSession.cpp:
```diff
@@ -223,19 +223,22 @@ void NetworkSession::destroyPrivateClickMeasurementStore(CompletionHandler<void(
+void NetworkSession::setTrackingPreventionEnabled(bool enabled)
 {
     ASSERT(!m_isInvalidated);
+    bool currentEnabled = !!m_resourceLoadStatistics;
```
`currentEnabled` sounds weird in English and boolean variables need a prefix as per WebKit coding style. I suggest `isCurrentlyEnabled`

cdumez commented on Source/WebKit/NetworkProcess/NetworkSession.cpp:
```diff
@@ -223,19 +223,22 @@ void NetworkSession::destroyPrivateClickMeasurementStore(CompletionHandler<void(
+    if (currentEnabled == enabled)
+        return;
+
+    RELEASE_LOG(Storage, "%p - NetworkSession::setTrackingPreventionEnabled sessionID=%" PRIu64 ", enabled=%d", this, m_sessionID.toUInt64(), enabled);
```
I think we usually have a `:` after the function name in our release logging.

cdumez reviewed (approved):
LGTM with nit.

szewai committed:
Ensure TrackingPreventionEnabled flag is correctly set before using WebsiteDataStore
https://bugs.webkit.org/show_bug.cgi?id=270382
rdar://123922825

Reviewed by Chris Dumez.

We currently set TrackingPreventionEnabled on WebsiteDataStore in WebProcessPool::createNewWebProcess. This has caused
problem that TrackingPreventionEnabled is incorrectly set when WebsiteDataStore (or its corresponding NetworkSession)
is in use. One case is when process is created for prewarm, it does not have a attached WebsiteDataStore (it will be
assigned a WebsiteDataStore later), so we don't set TrackingPreventionEnabled flag, and we don't set it when the
prewarmed process is used for loading (WebsiteDataStore is assigned to it). Another case is when WebsiteDataStore is
used without web process for website data retrieval. Failing to set the flag will cause ITP data not to be fetched or
removed.

To fix this issue, this patch moves TrackingPreventionEnabled setting to WebsiteDataStore: WebsiteDataStore will get
the right value when it needs to send its parameters to network process and web process (i.e. it's going to be used).
To simply the setting logic, this patch also makes the following changes:
1. WebsiteDataStore::m_trackingPreventionEnabled now has 3 states: yes, no and default, where default means we will fall
back to TCC permission state to decide the value. This enables us to remove m_isTrackingPreventionStateExplicitlySet.
2. Drop appHasRequestedCrossWebsiteTrackingPermission and TrackingPreventionStateExplicitlySet parameters since network
process does not need to re-compute the TrackingPreventionEnabled value and know about how it is computed. We previously
used it for logging and debugging purpose, and we probably don't need it any more (as TCC check is not free). This patch
replaces it with normal logging in NetworkSession::setTrackingPreventionEnabled.

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::setTrackingPreventionEnabled):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.serialization.in:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::NetworkSessionCocoa):
(WebKit::activateSessionCleanup): Deleted.
* Source/WebKit/Shared/ResourceLoadStatisticsParameters.h:
* Source/WebKit/Shared/ResourceLoadStatisticsParameters.serialization.in:
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.cpp:
(WKWebsiteDataStoreSetResourceLoadStatisticsEnabled):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _setResourceLoadStatisticsEnabled:]):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createNewWebProcess):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::WebsiteDataStore):
(WebKit::WebsiteDataStore::setIsRunningResourceLoadStatisticsTest):
(WebKit::WebsiteDataStore::defaultTrackingPreventionEnabled const):
(WebKit::WebsiteDataStore::trackingPreventionEnabled const):
(WebKit::WebsiteDataStore::setTrackingPreventionEnabled):
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::isTrackingPreventionStateExplicitlySet const): Deleted.
(WebKit::WebsiteDataStore::useExplicitTrackingPreventionState): Deleted.

Canonical link: https://commits.webkit.org/275635@main
